GitExtUtils/GitArgumentBuilder.cs

```diff
@@ -0,0 +1,199 @@
+    /// };
+    /// </code>
+    /// </example>
+    public sealed class GitArgumentBuilder : ArgumentBuilder
+    {
+        private static readonly Regex _commandRegex = new Regex("^[a-z0-9_.-]+$", RegexOptions.Compiled);
+
+        private readonly List<GitConfigItem> _configItems;
+        private readonly string _command;
+
+        /// <summary>
+        /// Initialises a new <see cref="GitArgumentBuilder"/> for the given <paramref name="command"/>.
+        /// </summary>
+        /// <param name="command">The git command this builder is compiling arguments for.</param>
+        /// <param name="commandConfiguration">Optional source for default command configuration items. Pass <c>null</c> to use the Git Extensions defaults.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="command"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentException"><paramref name="command"/> is an invalid string.</exception>
+        public GitArgumentBuilder([NotNull] string command, [CanBeNull] GitCommandConfiguration commandConfiguration = null)
+        {
+            if (command == null)
+            {
+                throw new ArgumentNullException(nameof(command));
+            }
+
+            if (!_commandRegex.IsMatch(command))
+            {
+                throw new ArgumentException($"Git command \"{command}\" contains invalid characters.", nameof(command));
+            }
+
+            _command = command;
+
+            commandConfiguration = commandConfiguration ?? GitCommandConfiguration.Default;
+
+            var defaultConfig = commandConfiguration.Get(command);
+            _configItems = new List<GitConfigItem>(capacity: defaultConfig.Count + 2);
+            if (defaultConfig.Count != 0)
+            {
+                _configItems.AddRange(defaultConfig);
+            }
+        }
+
+        /// <summary>
+        /// Add <paramref name="configItem"/> to this builder.
+        /// </summary>
+        /// <remarks>
+        /// Any prior config item with the same key will be replaced.
+        /// </remarks>
+        /// <param name="configItem">The config item to add to the builder.</param>
+        public void Add(GitConfigItem configItem)
+        {
+            // Append or replace config item based upon its key
+            var index = _configItems.IndexOf(item => string.Equals(item.Key, configItem.Key, StringComparison.OrdinalIgnoreCase));
+
+            if (index == -1)
+            {
+                _configItems.Add(configItem);
+            }
+            else
+            {
+                _configItems[index] = configItem;
+            }
+        }
+
+        /// <inheritdoc />
+        public override string ToString()
+        {
+            var arguments = base.ToString();
+
+            var capacity = _command.Length + 1 + arguments.Length + 1 + _configItems.Sum(i => i.Key.Length + i.Value.Length + 5);
```

Comment:
+7 instead of 5 because of possible quotes?

Reply:
Thanks, updated, with comment to explain magic number.